src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,130 @@
+import * as React from "react";
+import AsyncStorage from "@react-native-async-storage/async-storage";
+
+import { PASSWORD_KEY } from "../constants";
+
+type AppModeTypes = "normal" | "kill";
+type AuthStatusTypes = "pending" | "authenticated" | "notRequired";
+
+type AuthState = {
+  passcode: string | null;
+  appMode: AppModeTypes;
+  killModeEnabled: boolean;
+  authStatus: AuthStatusTypes;
+};
+
+type KillModeActions =
+  | { type: "setPasscode"; newPasscode: string | null }
+  | { type: "toggleAppMode"; newAppMode?: AppModeTypes }
+  | { type: "toggleKillModeEnabled"; newKillModeValue?: boolean }
+  | { type: "setAuthStatus"; newAuthStatus: AuthStatusTypes };
+
+function killModeReducer(state: AuthState, action: KillModeActions): AuthState {
+  switch (action.type) {
+    /***
+     * if passcode is set to null we need to set the auth status to 'notRequired'
+     *
+     * If a passcode is set, we want to set the authStatus to 'authenticated'
+     * If the auth status is set to 'pending' they will be navigated to the auth screen, to type in their password.
+     */
+    case "setPasscode":
+      const passcode = action.newPasscode;
+      if (passcode === null || validPasscode(passcode)) {
+        AsyncStorage.setItem(PASSWORD_KEY, passcode || "");
+        if (!passcode) {
+          return { ...state, passcode, authStatus: "notRequired" };
+        }
+        return { ...state, passcode, authStatus: "authenticated" };
+      }
+      throw new Error("Invalid New Password");
+    /***
+     * Does not require a new value in the dispatcher payload, but a new value is optional in the payload
+     * If there is no new value, the value is toggled (as there are only 2 values)
+     */
+    case "toggleAppMode":
+      return {
+        ...state,
+        appMode: !!action.newAppMode
+          ? action.newAppMode
+          : state.appMode === "kill"
+          ? "normal"
+          : "kill",
+      };
```

Comment:
nit: i find the ternary a little hard to parse on quick reads here. definitely harder if we ever end up with more than two valid app modes. without getting into state machines, maybe something like this would be clearer despite being slightly more verbose:

```ts
let mode: AppModeTypes;

if (action.newAppMode) {
  mode = action.newAppMode;
} else {
  switch (state.appMode) {
    case "kill":
      mode = "normal";
      break;
    case "normal":
      mode = "kill";
      break;
  }
}

return {
  ...state,
  appMode: mode,
};
```

Reply:
think the better suggestion is outlined in this [comment](https://github.com/digidem/mapeo-mobile/pull/923/files#r810353052) i.e. create separate actions + reducers for toggling vs setting state